File: tests/algorithms/x/Java/electronics/carrier_concentration.java
public class Main {
    static class CarrierResult {
        String name;
        double value;
        CarrierResult(String name, double value) {
            this.name = name;
            this.value = value;
        }
        CarrierResult() {}
        @Override public String toString() {
            return String.format("{'name': '%s', 'value': %s}", String.valueOf(name), String.valueOf(value));
        }
    }

    static CarrierResult r1;
    static CarrierResult r2;
    static CarrierResult r3;

    static double sqrtApprox(double x) {
        double guess = (double)((double)(x) / (double)(2.0));
        long i_1 = 0L;
        while ((long)(i_1) < 20L) {
            guess = (double)((double)(((double)(guess) + (double)((double)(x) / (double)(guess)))) / (double)(2.0));
            i_1 = (long)((long)(i_1) + 1L);
        }
        return guess;
    }

    static CarrierResult carrier_concentration(double electron_conc, double hole_conc, double intrinsic_conc) {
        long zero_count = 0L;
        if ((double)(electron_conc) == (double)(0.0)) {
            zero_count = (long)((long)(zero_count) + 1L);
        }
        if ((double)(hole_conc) == (double)(0.0)) {
            zero_count = (long)((long)(zero_count) + 1L);
        }
        if ((double)(intrinsic_conc) == (double)(0.0)) {
            zero_count = (long)((long)(zero_count) + 1L);
        }
        if ((long)(zero_count) != 1L) {
            throw new RuntimeException(String.valueOf("You cannot supply more or less than 2 values"));
        }
        if ((double)(electron_conc) < (double)(0.0)) {
            throw new RuntimeException(String.valueOf("Electron concentration cannot be negative in a semiconductor"));
        }
        if ((double)(hole_conc) < (double)(0.0)) {
            throw new RuntimeException(String.valueOf("Hole concentration cannot be negative in a semiconductor"));
        }
        if ((double)(intrinsic_conc) < (double)(0.0)) {
            throw new RuntimeException(String.valueOf("Intrinsic concentration cannot be negative in a semiconductor"));
        }
        if ((double)(electron_conc) == (double)(0.0)) {
            return new CarrierResult("electron_conc", (double)(((double)(intrinsic_conc) * (double)(intrinsic_conc))) / (double)(hole_conc));
        }
        if ((double)(hole_conc) == (double)(0.0)) {
            return new CarrierResult("hole_conc", (double)(((double)(intrinsic_conc) * (double)(intrinsic_conc))) / (double)(electron_conc));
        }
        if ((double)(intrinsic_conc) == (double)(0.0)) {
            return new CarrierResult("intrinsic_conc", sqrtApprox((double)((double)(electron_conc) * (double)(hole_conc))));
        }
        return new CarrierResult("", -1.0);
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            r1 = carrier_concentration((double)(25.0), (double)(100.0), (double)(0.0));
            System.out.println(r1.name + ", " + _p(r1.value));
            r2 = carrier_concentration((double)(0.0), (double)(1600.0), (double)(200.0));
            System.out.println(r2.name + ", " + _p(r2.value));
            r3 = carrier_concentration((double)(1000.0), (double)(0.0), (double)(1200.0));
            System.out.println(r3.name + ", " + _p(r3.value));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Pascal/electronics/carrier_concentration.pas
{$mode objfpc}
program Main;
uses SysUtils;
type CarrierResult = record
  name: string;
  value: real;
end;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  r1: CarrierResult;
  r2: CarrierResult;
  r3: CarrierResult;
  intrinsic_conc: real;
  hole_conc: real;
  x: real;
  electron_conc: real;
function makeCarrierResult(name: string; value: real): CarrierResult; forward;
function sqrtApprox(x: real): real; forward;
function carrier_concentration(electron_conc: real; hole_conc: real; intrinsic_conc: real): CarrierResult; forward;
function makeCarrierResult(name: string; value: real): CarrierResult;
begin
  Result.name := name;
  Result.value := value;
end;
function sqrtApprox(x: real): real;
var
  sqrtApprox_guess: real;
  sqrtApprox_i: integer;
begin
  sqrtApprox_guess := x / 2;
  sqrtApprox_i := 0;
  while sqrtApprox_i < 20 do begin
  sqrtApprox_guess := (sqrtApprox_guess + (x / sqrtApprox_guess)) / 2;
  sqrtApprox_i := sqrtApprox_i + 1;
end;
  exit(sqrtApprox_guess);
end;
function carrier_concentration(electron_conc: real; hole_conc: real; intrinsic_conc: real): CarrierResult;
var
  carrier_concentration_zero_count: integer;
begin
  carrier_concentration_zero_count := 0;
  if electron_conc = 0 then begin
  carrier_concentration_zero_count := carrier_concentration_zero_count + 1;
end;
  if hole_conc = 0 then begin
  carrier_concentration_zero_count := carrier_concentration_zero_count + 1;
end;
  if intrinsic_conc = 0 then begin
  carrier_concentration_zero_count := carrier_concentration_zero_count + 1;
end;
  if carrier_concentration_zero_count <> 1 then begin
  panic('You cannot supply more or less than 2 values');
end;
  if electron_conc < 0 then begin
  panic('Electron concentration cannot be negative in a semiconductor');
end;
  if hole_conc < 0 then begin
  panic('Hole concentration cannot be negative in a semiconductor');
end;
  if intrinsic_conc < 0 then begin
  panic('Intrinsic concentration cannot be negative in a semiconductor');
end;
  if electron_conc = 0 then begin
  exit(makeCarrierResult('electron_conc', (intrinsic_conc * intrinsic_conc) / hole_conc));
end;
  if hole_conc = 0 then begin
  exit(makeCarrierResult('hole_conc', (intrinsic_conc * intrinsic_conc) / electron_conc));
end;
  if intrinsic_conc = 0 then begin
  exit(makeCarrierResult('intrinsic_conc', sqrtApprox(electron_conc * hole_conc)));
end;
  exit(makeCarrierResult('', -1));
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  r1 := carrier_concentration(25, 100, 0);
  writeln((r1.name + ', ') + FloatToStr(r1.value));
  r2 := carrier_concentration(0, 1600, 200);
  writeln((r2.name + ', ') + FloatToStr(r2.value));
  r3 := carrier_concentration(1000, 0, 1200);
  writeln((r3.name + ', ') + FloatToStr(r3.value));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
